### PR TITLE
Add delete text action

### PR DESCRIPTION
**Pull request by yibolu**
Added delete text action and its test.
The delete text action will be used by the BACKSPACE handler, which will delete a single char if a single place gets selected, and will delete a chunk of text if they are selected.

Currently, delete action has two modes: 
1.  delete single char
2.  delete multiple chars among either multiple segments/single segment.

Only #1 will be used currently since we don't have the handlers for mouse up/down, and selection service either.

### CODE REVIEW

File: src/common/array.ts [hunk 1]
```diff
@@ -1,3 +1,10 @@
 export function insert<T>(arr: T[], index: number, slice: T[]): T[] {
     return [...arr.slice(0, index), ...slice, ...arr.slice(index)];
 }
+
+export function remove<T>(arr: T[], idx: number, count: number = 1): T[] {
+    return [
+        ...arr.slice(0, idx),
+        ...arr.slice(idx + count)
+    ];
+}
```
Comment (naturetime10) [inline src/common/array.ts:10]: We are going to use those helpers over & over again. Let's include unit tests for them so that we are confident they are working correctly when dealing with complexities for the actions.

File: src/state/delete-text.action.test.ts [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+                cursor: {
+                    startSegmentIndex: 0,
+                    startOffset: 1,
+                    endSegmentIndex: 0,
+                    endOffset: 1
+                }
+            }
+        },
+        {
+            name: 'should delete 1 character at cursor position and should delete the segment when segment is empty',
```
Comment (naturetime10) [inline src/state/delete-text.action.test.ts:64]:
```suggestion
            name: 'should delete 1 character and the empty segment',
```

File: src/state/delete-text.action.test.ts [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+                cursor: {
+                    startSegmentIndex: 0,
+                    startOffset: 3,
+                    endSegmentIndex: 0,
+                    endOffset: 3
+                }
+            }
+        },
+        {
+            name: 'should delete multiple characters at cursor selection',
```
Comment (naturetime10) [inline src/state/delete-text.action.test.ts:119]:
```suggestion
            name: 'should delete all selected characters in a single segment',
```

File: src/state/delete-text.action.test.ts [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+describe('DeleteTextAction', () => {
+    interface ITestCase {
+        name: string;
+        currState: IEditorState;
+        expectedNewState: IEditorState;
+    }
+
+    const testCases: ITestCase[] = [
+        {
+            name: 'should delete 1 character at cursor position',
```
Comment (naturetime10) [inline src/state/delete-text.action.test.ts:15]:
```suggestion
            name: 'should delete 1 character in a single segment',
```

File: src/state/delete-text.action.test.ts [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+                cursor: {
+                    startSegmentIndex: 1,
+                    startOffset: 3,
+                    endSegmentIndex: 1,
+                    endOffset: 3
+                }
+            }
+        },
+        {
+            name: 'should delete multiple characters at cursor selection among different start/end segment',
```
Comment (naturetime10) [inline src/state/delete-text.action.test.ts:168]:
```suggestion
            name: 'should delete multiple selected characters across 2 segments',
```

File: src/state/delete-text.action.test.ts [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+                cursor: {
+                    startSegmentIndex: 0,
+                    startOffset: 2,
+                    endSegmentIndex: 0,
+                    endOffset: 2
+                }
+            }
+        },
+        {
+            name: 'should delete text among more than 2 segments and have correct index',
```
Comment (naturetime10) [inline src/state/delete-text.action.test.ts:217]:
```suggestion
            name: 'should delete text across more than 2 segments with correct final index',
```

File: src/state/delete-text.action.test.ts [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+                cursor: {
+                    startSegmentIndex: 0,
+                    startOffset: 2,
+                    endSegmentIndex: 0,
+                    endOffset: 2
+                }
+            }
+        },
+        {
+            name: 'edge case: test when delete all content among three segments',
```
Comment (naturetime10) [inline src/state/delete-text.action.test.ts:272]:
```suggestion
            name: 'delete all content across three segments',
```

File: src/state/delete-text.action.test.ts [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+                    startOffset: 0,
+                    endSegmentIndex: 2,
+                    endOffset: 4
+                }
+            },
+            expectedNewState: {
+                options: [],
+                segments: [
+                ],
+                cursor: {
```
Comment (naturetime10) [inline src/state/delete-text.action.test.ts:306]: Should we have an empty segment here with cursor moved to the start? This will also work well with the following cursor position:

```
startSegmentIndex: 0,
startOffset: 0,
endSegmentIndex: 0,
endOffset: 0
```

File: src/state/delete-text.action.test.ts [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+                    startSegmentIndex: 0,
+                    startOffset: 0,
+                    endSegmentIndex: 0,
+                    endOffset: 0
+                }
+            }
+        },
+        {
+            name: 'edge case: test when end segment\'s content is removed',
```
Comment (naturetime10) [inline src/state/delete-text.action.test.ts:315]: cursor is at the correct location when the middle segment is deleted completely

File: src/state/delete-text.action.ts [hunk 10]
```diff
@@ -0,0 +1,185 @@
+import { Action } from './action';
+import { ISegment } from '../entities/segment';
+import { ICursor } from '../entities/cursor';
+import { getCurrSegment, IEditorState, findSegment } from './editor.state';
+import { remove } from '../common/array';
+
+export class DeleteTextAction implements Action {
+
```
Comment (naturetime10) [inline src/state/delete-text.action.ts:8]:
```suggestion
```

File: src/state/delete-text.action.ts [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+            endOffset: newCursor.startOffset
+        });
+
+        return Object.assign({}, state, {
+            segments: newSegments,
+            cursor: newCursor
+        });
+    }
+
+    public perform(state: IEditorState): IEditorState {
```
Comment (naturetime10) [inline src/state/delete-text.action.ts:168]: Let's move public function to the top so that it's easier for readers to follow the logic.

File: src/state/delete-text.action.ts [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+        const { cursor, segments } = state;
+
+        if (segments.length < 1) {
+            return state;
+        }
+
+        if (cursor.startSegmentIndex >= segments.length || cursor.endSegmentIndex < 0 || cursor.startSegmentIndex > cursor.endSegmentIndex) {
```
Comment (naturetime10) [inline src/state/delete-text.action.ts:175]: hasValidSegmentIDs(cursor)

Comment (naturetime10) [reply src/state/delete-text.action.ts:175]: Let's put the long line into `hasValidSegmentIDs` to simplify the code.

File: src/state/delete-text.action.ts [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+            return state;
+        }
+
+        if (cursor.startSegmentIndex === cursor.endSegmentIndex && cursor.startOffset === cursor.endOffset) {
+            return this.deleteSingleChar(state);
+        } else {
+            return this.deleteMultiple(state);
+        }
```
Comment (naturetime10) [inline src/state/delete-text.action.ts:183]:
```suggestion
      return this.deleteMultiple(state);
```

File: src/state/delete-text.action.ts [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+        });
+    }
+
+    private deleteSingleChar(state: IEditorState): IEditorState {
+        const { cursor, segments } = state;
+        let newState: IEditorState = state;
+        let newSegments: ISegment[] = segments;
+        let newCursor: ICursor = cursor;
+
+        const segment = getCurrSegment(newState);
```
Comment (naturetime10) [inline src/state/delete-text.action.ts:27]:
```suggestion
        const segment = getCurrSegment(state);
```

File: src/state/delete-text.action.ts [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+        if (!segment) {
+            return state;
+        }
+
+        const newSegment = Object.assign<any, ISegment, Partial<ISegment>>(
+            {},
+            segment,
+            {
+                // to remove the single char at the cursor
```
Comment (naturetime10) [inline src/state/delete-text.action.ts:36]:
```suggestion
                // cursor is always located at the end of the previous segment when visually it should be located in the middle of 2 segments
```

File: src/state/delete-text.action.ts [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+                content: remove<string>(
+                    segment.content,
+                    newCursor.startOffset - 1
```
Comment (naturetime10) [inline src/state/delete-text.action.ts:39]:
```suggestion
                    segment.content,
                    newCursor.startOffset - 1,
                    1
```
Prefer making thing explicit to avoid confusion.

File: src/state/delete-text.action.ts [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+                    endSegmentIndex: 0,
+                    endOffset: 0
+                });
+            }
+            
+        } else {
+            newCursor = Object.assign({}, newCursor, {
+                startOffset: newCursor.startOffset - 1,
+                endOffset: newCursor.endOffset - 1
+            });
```
Comment (naturetime10) [inline src/state/delete-text.action.ts:83]: Let's handle this case first in the `if` statement.

File: src/state/delete-text.action.ts [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+
+            // If it's the first segment, we need to reset the cursor.
+            } else {
+                newCursor = Object.assign({}, newCursor, {
+                    startSegmentIndex: 0,
+                    startOffset: 0,
+                    endSegmentIndex: 0,
+                    endOffset: 0
+                });
+            }
```
Comment (naturetime10) [inline src/state/delete-text.action.ts:77]:
```suggestion
           {newSegments, newCursor} = this.deleteEmptySegment(newSegment);
```

Comment (naturetime10) [reply src/state/delete-text.action.ts:77]: Let's move this inside `deleteEmptySegment`. This will reduce the number of nest `if else` statement in this function to simplify the logic. Inside `deleteEmptySegment`, we then only need one `if` statement to deal with `cursor` jumping.

File: src/state/delete-text.action.ts [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+        let endSegment = findSegment(cursor.endSegmentIndex, segments);
+        if (!endSegment) {
+            return state;
+        }
+
+        startSegment = Object.assign<any, ISegment, Partial<ISegment>>(
+            {},
+            startSegment,
+            {
+                // to remove the single char at the cursor
```
Comment (naturetime10) [inline src/state/delete-text.action.ts:111]:
```suggestion
```
We are removing multiple characters here.

File: src/state/delete-text.action.ts [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+                    startSegment.content.length - cursor.startOffset
+                )
+            }
+        );
+
+        endSegment = Object.assign<any, ISegment, Partial<ISegment>>(
+            {},
+            endSegment,
+            {
+                // to remove the single char at the cursor
```
Comment (naturetime10) [inline src/state/delete-text.action.ts:124]:
```suggestion
```

File: src/state/delete-text.action.ts [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+            if (currSegment.index === endSegment.index) {
+                return endSegment;
+            }
+            return currSegment;
+        });
+
+        let startIndex = startSegment.index;
+        let endIndex = endSegment.index;
+        if (startSegment.content.length === 0) {
+            startIndex = startIndex - 1;
```
Comment (naturetime10) [inline src/state/delete-text.action.ts:146]:
```suggestion
            startIndex--;
```

File: src/state/delete-text.action.ts [hunk 22]
```diff
@@ -0,0 +1,185 @@
...
+        }
+        if (endSegment.content.length === 0) {
+            endIndex = endIndex + 1;
```
Comment (naturetime10) [inline src/state/delete-text.action.ts:149]:
```suggestion
            endIndex++;
```

File: src/state/delete-text.action.ts [hunk 23]
```diff
@@ -0,0 +1,185 @@
...
+        }
+
+        if (endIndex - startIndex - 1 > 0) {
+            newSegments = remove<ISegment>(newSegments, startIndex + 1, endIndex - startIndex - 1);
```
Comment (naturetime10) [inline src/state/delete-text.action.ts:153]: I am curious about the `-1` here. How does it work when the 2 segments are adjacent? Or when only the first segment is deleted?

File: src/state/delete-text.action.ts [hunk 24]
```diff
@@ -0,0 +1,185 @@
...
+            newSegments = this.rematchIndex(newSegments);
```
Comment (naturetime10) [inline src/state/delete-text.action.ts:154]:
```suggestion
            newSegments = this.reassignIndices(newSegments);
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: src/state/delete-text.action.ts (hunk 23 above)
Comment (naturetime10) [reply src/state/delete-text.action.ts:153]: Resolved during zoom meeting. Misread a the offset.
